### PR TITLE
Add RoundChangePayload validation for QBFT

### PR DESCRIPTION
**Pull request by rain-on**
Signed-off-by: Trent Mohay <trent.mohay@consensys.net>

## PR description
Adds the SignedData<RoundChangePayload> validation code, along with unit tests.
Does not wire this validation code into the QBFT business logic at this stage.
## Fixed Issue(s)
N/A

## Changelog

- [X] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

**Review by usmansaleem — APPROVED** [review]
LGTM

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangePayloadValidator.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      LOG.info("{}: did not target expected height", ERROR_PREFIX);
+      return false;
+    }
+
+    final int targetRound = payload.getRoundIdentifier().getRoundNumber();
+    if (targetRound <= 0) {
+      LOG.info("{}: must contain a positive target round number", ERROR_PREFIX);
+      return false;
+    }
+
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangePayloadValidator.java:64]: The QBFT spec also requires this check
```
|seqToSet(prepares)| >= quorum(|validators|)
```

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangePayloadValidator.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      if (metadata.getPreparedRound() >= targetRound) {
+        LOG.info("{}: prepared metadata is from a round ahead of target round", ERROR_PREFIX);
+        return false;
+      }
+
+      if (metadata.getPreparedRound() < 0) {
+        LOG.info("{}: prepared metadata is from a negative round number", ERROR_PREFIX);
+        return false;
+      }
+    }
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangePayloadValidator.java:76]: We also need to check that the digest(proposalForNextRound) === prepareValue

From the QBFT dafny spec:
```
digest(proposalForNextRound) == preparedValue
```

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/ValidationHelpers.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+import org.hyperledger.besu.consensus.common.bft.payload.Authored;
+import org.hyperledger.besu.consensus.common.bft.payload.Payload;
+import org.hyperledger.besu.consensus.common.bft.payload.SignedData;
+
+import java.util.Collection;
+
+public class ValidationHelpers {
+
+  public static <T extends Payload> boolean hasDuplicateAuthors(
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/ValidationHelpers.java:25]: These could be in BFT common code

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangePayloadValidator.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      return false;
+    }
+
+    if (payload.getPreparedRoundMetadata().isPresent()) {
+      final PreparedRoundMetadata metadata = payload.getPreparedRoundMetadata().get();
+      if (metadata.getPreparedRound() >= targetRound) {
+        LOG.info("{}: prepared metadata is from a round ahead of target round", ERROR_PREFIX);
+        return false;
+      }
+
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangePayloadValidator.java:71]: My understanding is that if we have a preparedValue and preparedRound we must also have the proposalForNextRound. So proposedBlock must not be empty. Would be this in the RoundChangeMessageValidator?

From the dafny QBFT spec:
```
            && optionIsPresent(uPayload.preparedValue)
            && optionIsPresent(uPayload.preparedRound)
            && optionIsPresent(proposalForNextRound)
```

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangePayloadValidator.java (hunk 1 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangePayloadValidator.java:64]: Doesn't belong at this level - this is purely to validate JUST the RoundChangePayload (not message) which doesn't contain the prepare msgs.

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangePayloadValidator.java (hunk 2 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangePayloadValidator.java:76]: Again - can't be checked at this level, must be done at the message level.

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangePayloadValidator.java (hunk 4 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangePayloadValidator.java:71]: This level of test cannot check that, as the block is not contained within the RoundChangePayload

**Review by jframe — APPROVED** [review]

**Pull request merged**